### ISSUE TITLE
Question on BTC fees.

### ISSUE THREAD

**Issue by Crypto2** [OmniLayer/omnicore#383]
In an omnicored wallet if address X has say MAID coins and address Y has some Bitcoin; when I use omni_send will it use the coins in address Y to pay the Bitcoin TX fee or does it have to have coins in address X? If so, what is the recommended procedure as a service provider when you assign user ABC a deposit address? Delay withdrawals by sending some Bitcoin to the address and waiting for a confirm before performing the omni_send? For payment processing I could see it being even worse since the address would only be used once, it seems like there would always be waste since you would have to add enough Bitcoin to cover the TX fee without knowing exactly how much the fee would be.

**Comment by zathras-crypto** [comment]
Hey @Crypto2 

> In an omnicored wallet if address X has say MAID coins and address Y has some Bitcoin; when I use omni_send will it use the coins in address Y to pay the Bitcoin TX fee or does it have to have coins in address X?

Omni Core will not use a different address for fees if the sending address has no BTC I'm afraid - Omni identifies the sending address by the inputs, so if we only used the inputs from address Y then the Omni transaction will be parsed as being sent from address Y (and not address X).

> If so, what is the recommended procedure as a service provider when you assign user ABC a deposit address?

This is an interesting question, and if you don't mind I'll get a little technical on exactly what we're doing in this regard...

A little while ago we activated a new class of transaction (class C) which moves away from our old encoding mechanisms to instead use nulldata (OP_RETURN) outputs.  As part of this transaction class we also updated the sender handling so that the first VIN is always considered the sender, and any remaining inputs are ignored (from a sender identification perspective).

As such with a Class C transaction, only the first input has to come from the sender, and the rest can be from any other address. 

When we apply that logic to the service provider deposit/sweep/withdrawal consideration, we can see that:

1) A user sends some tokens to ```DEPOSITADDRESS```.  As part of this (and any Omni) send, a small amount of BTC (a touch over the dust threshold) is also sent to ```DEPOSITADDRESS``` for recipient identification.

2) Service provider attaches the output from the users' send in step 1 as the first input to a sweep/withdrawal transaction thus identifying ```DEPOSITADDRESS``` as the sender.  

3) Service provider attaches outputs from any other transaction/address as additional inputs to this sweep/withdrawal transaction to pay the remainder of fees.

The idea here being that we can always use fees from other addresses as long as there is at least one input to attach to the transaction as first VIN (which there is if the address has received a send).

Hope that helps provide a bit of a base, and please don't hesitate to ask any followup questions.

Thanks :)
Z

**Comment by zathras-crypto** [comment]
I should also note that this is considered advanced usage and not currently done natively / automatically by the Omni Core wallet tx builder (eg ```omni_send``` and friends) but can be used via the raw tx interface (```omni_createraw``` commands).

**Comment by Crypto2** [comment]
Are there any plans on adding that to Omni Core so it does it automatically? Possibly with a new RPC omni_send parameter or omnicored -option to not break backwards compatibility.

Of course I'm just a johnny come lately and only have a cursory understanding on the Omni Layer technical details, I was just thinking as an option instead of using the vin to determine the sender the Bitcoin address could be included in the the data (20 or 21 bytes I think) and signed by that address as proof of ownership. I guess this would substantially increase the amount of data embedded in the blockchain though, it has probably already been thought of and rejected for that reason.

**Comment by zathras-crypto** [comment]
> Are there any plans on adding that to Omni Core so it does it automatically? Possibly with a new RPC omni_send parameter or omnicored -option to not break backwards compatibility. It seems like it would help a lot of people and make entering the Omni world easier.

There are indeed plans to add this functionality to Omni Core, but as with most projects it all comes down to prioritization.  This is functionality that would help integrators I completely agree and I'm very supportive of adding it sooner rather than later.

> .....I was just thinking as an option instead of using the vin to determine the sender the Bitcoin address could be included in the the data (20 or 21 bytes I think) and signed by that address as proof of ownership.....

A signature is roughly 70 bytes (give or take) so in addition to the HASH160 of the address we're at 90+ bytes just for the sender identification.  Nulldata (OP_RETURN) outputs are currently limited to 80 bytes via default Bitcoin Core policy so it's probably not a strategy we'd look at, but thanks for the suggestion!

**Comment by bumbar1** [comment]
@zathras-crypto could you provide an example on how to populate the `rawtransaction` field for `omni_sendrawtx` method. I'm guessing one would use `omni_createrawtx_opreturn` with `rawtx = null`(?), but what should go in `payload` field?

**Comment by zathras-crypto** [comment]
> could you provide an example on how to populate the rawtransaction field for omni_sendrawtx method. I'm guessing one would use omni_createrawtx_opreturn with rawtx = null(?)

@dexX7 I'm sure I've seen gist workups from you with examples for the raw transactions interface - am I remembering correctly?  If so can you please link for @bumbar1 ?  Otherwise we should do a guide for this.

> ...but what should go in payload field?

There are ```omni_createpayload``` calls but they hadn't been documented yet.  Here you go https://github.com/OmniLayer/omnicore/pull/385

**Comment by Crypto2** [comment]
I'm working up a patch right now, it's not super pretty but it's a starting point/proof of concept doing basically what you were talking about before. Basically was have one address, the Funding Address (FA) that holds BTC to pay the TX fees for all the transactions. The reason for a specific funding address is so it doesn't use inputs from addresses that have Omni Layer coins randomly and mess up sends from those later.

1. Modified the Omni SelectCoins function so it grabs only one input from the Sending Address (SA) and grabs anything else it needs from the funding address. (To minimize waste of SA inputs for later sends.)
2. I changed Coin Control to use a vector instead of set to maintain the order of the inputs, then in CreateTransaction it makes sure that the 1st item in the vector is the 1st in the new transaction.
3. Also in CreateTransaction, if there is more than 0.00005 BTC change it returns it to the FA and gives the small remainder back to the SA so if there is another send later it has an input to use.

Is there any importance to the order/amount of the outputs in the Bitcoin transaction? I know by default Bitcoin places the change randomly and that doesn't seem to have been altered.

Here is an example successful test with 12JYmnfYU2ghzjwUAspzJsSnmJtK9bZPYR as the FA:
https://blockchain.info/tx/2bf9592c84ec6c76a8cbeaf620de0545e6afe6fce046c3607daf1bbba0f0bf4c
http://omnichest.info/lookuptx.aspx?txid=2bf9592c84ec6c76a8cbeaf620de0545e6afe6fce046c3607daf1bbba0f0bf4c

**Comment by zathras-crypto** [comment]
> I'm working up a patch right now

Awesome!!! I'm impressed how quickly you've got up to speed! Thanks for contributing! :)

The model seems good and I like the idea of an explicit funding address (I'm all for explicit behaviour).  We could get away without it by first evaluating the tally when selecting inputs to check if an address holds Omni properties & if not, consider it safe to use for an input - though I'm +1 on the explicit address if @dexX7 you're cool with that too?

The functionality could be exposed via a configuration call, for example:
```omni_setaltfeesource false``` configures to use original model (default)
```omni_setaltfeesource true 1BitcoinAddressToUseForFees``` configures to use new model with specified FA

I'd be very interested to see what you've done when you're happy to share :)

> Is there any importance to the order/amount of the outputs in the Bitcoin transaction? I know by default Bitcoin places the change randomly and that doesn't seem to have been altered.

Actually yes - reference identification (the recipient) removes outputs to the sender as change but since the change output to the FA is not from the sender, it likely won't be detected as change.  The fallback in reference identification is always highest vout, so best bet is to always put the reference address as the last vout and then multiple change addresses shouldn't be an issue.

> Here is an example successful test

That's fantastic :)  Though if I may give a quick suggestion - create yourself a property in the test ecosystem (or ask one of us to send you some Test Omni) and use those for your test transactions - even though it's unlikely I'd hate to see you lose any funds while testing different ways of building transactions etc.

**Comment by Crypto2** [comment]
Yeah, if someone wants to send a few Test Omni to 1GP8Stw1G6bX4AG55fayx8DoKxr71jFBaK that would be cool.

I updated it now so it will always put the FA and SA change(s) right after the payload so the reference address will always be last: https://blockchain.info/tx/ec3726be944e84ea8212f12b6a5e4f9560257571d7ba29de706637372d356b0b

I created a PR at https://github.com/OmniLayer/omnicore/pull/388 (mainly for easy readability of multiple commits.) Fair warning to any non-Omni devs: I've only sent a few test transactions and only used omni_send. Needs more testing and review before production use.

**Comment by dexX7** [comment]
Hey guys,

I'm really happy to see more action, and thanks a lot for your contribution @Crypto2!

I generally see three options:

1. Transactions are manually created raw, with a combination of [omni_createrawtx_input](https://github.com/OmniLayer/omnicore/blob/omnicore-0.0.10/src/omnicore/doc/rpc-api.md#omni_createrawtx_input), [omni_createrawtx_opreturn](https://github.com/OmniLayer/omnicore/blob/omnicore-0.0.10/src/omnicore/doc/rpc-api.md#omni_createrawtx_opreturn), [omni_createrawtx_reference](https://github.com/OmniLayer/omnicore/blob/omnicore-0.0.10/src/omnicore/doc/rpc-api.md#omni_createrawtx_reference), [omni_createrawtx_change](https://github.com/OmniLayer/omnicore/blob/omnicore-0.0.10/src/omnicore/doc/rpc-api.md#omni_createrawtx_change), and the new [payload creation calls](https://github.com/OmniLayer/omnicore/blob/omnicore-0.0.10/src/omnicore/doc/rpc-api.md#omni_createpayload_simplesend). This can be done right now, but it involves fiddling with raw transactions on a lower level. I can provide an example in the next days.

2. We add a new set of API calls to facilitate the external funding explicitly, e.g. in form of `omni_funded_send`, `omni_funded_xxx`, ... When using these calls, the transactions are created by Omni Core, and funded by a specified source.

3. We use a global funding source, which can be configured, as proposed in the earlier posts and #388.

I could imagine crafting all transactions manually could be a bit annoying, but I could also imagine having a global funding source might be too inflexible for some cases.

Given that this feature was also requested by Tether, it would be interesting to learn more about the specific use cases.

**Comment by Crypto2** [comment]
The main reason I went with #3 which required mods to Bitcoin's CreateTransaction was because you don't know what the fee is anymore until then. Although #2 could be implemented the same way, just a little more work since you'd have to add parameters to several functions instead of setting a global variable. Depending on how multithreaded your application is being able to change the funding address on the fly via RPC in #388 does add some flexibility, in the end having omni_funded_send may be the way to go though.

I know manually crafting every transaction is a pain I don't want to bear :)

**Comment by zathras-crypto** [comment]
Awesome, thanks for posting the code @Crypto2 :)

I think option 1 is a no-go - we definitely want a raw tx interface to allow people to get creative, but alternative funding for fees is something many folks ask for, so it makes sense to make it as painless as possible.

Since the funding address can be changed during runtime I don't see a huge lack of flexibility here, but it does preclude the use of multiple funding addresses at the same time which might be an issue.

I think I'd agree it'd be great to understand where Tether stand on this - @CraigSellars can you add anything here?

**Comment by zathras-crypto** [comment]
@Crypto2 - just checking in, don't want you to think we've forgotten about you :)

I had a fall & broke my arm so have been taking a few personal days...

@CraigSellars can you please feedback on Tether requirements?

**Comment by Crypto2** [comment]
So far so good with the code as-is; not exactly high volume but no problems so far.

**Comment by zathras-crypto** [comment]
@CraigSellars - ping - can you please provide feedback?  Thanks

**Comment by CraigSellars** [comment]
@zathras-crypto and @crypto2 this is quite fantastic.  We have a large integrator who needs this functionality (managing the deposits for a huge number of users is far too cumbersome).

I am in favor of a modification of #2 and #3 as described… a config option that allows you to specify a funding address (and uses it when necessary), but I don’t *think* we need a new API call - I was thinking an alternate paramater for funding address at the end of the list of attributes would suffice.

If we have a working version of this, there are integrators who would use it immediately - what is the current status of this patch?

**Comment by Crypto2** [comment]
As far as the status we haven't had any trouble so far. The only major edge case left is if there aren't any inputs from the sending address to use it will still fail even if there are plenty in the funding address, no real way around that unless we auto generate another TX to the sending address and then use it in the new transaction.

**Comment by Crypto2** [comment]
Quick update nearly a year on, my patch is still working great with no issues. Not sure why it was never merged and integrated?

**Comment by zathras-crypto** [comment]
Your timing is perfect @crypto2 :)

Firstly please allow me to apologize that this got forgotten about - we have a small development team and are often attacking many things at once which sadly can cause issues to lose focus.

I actually raised with the team not two weeks ago that baaed on feedback we needed to assist our integrators with two functions; central fee addressing and batch sends (multiple sends in a single transaction).  There was unanimous agreement to shift focus to these two items next.

I'm just finishing up a rewrite of sending in the UI for our Omni Lite prototype after which I'll be looking at these items (central fee/batch withdrawals) & will definitely start with your PR.

TL:DR; great to hear your patch is working & be back to you re refining/merging soon :)

**Comment by Crypto2** [comment]
Any updates on this? Unfortunately with the latest hardfork/mandatory update I'm having to try to update the patch. Seems to be working for the most part but isn't signing the TX properly somehow; haven't figured that out yet even though the signing code is untouched.

**Comment by Crypto2** [comment]
Actually, while this still would be nice to have in the core daemon I ended up just doing the same thing on my end with the payload functions. Took less than an hour to do the same thing versus hacking on the wallet :)

**Comment by dexX7** [comment]
Hey @Crypto2, I'm very sorry for breaking your implementation! How did you end up doing it? With the raw transaction build calls?

**Comment by Crypto2** [comment]
No worries :) Yep, I reimplemented with the raw transaction calls.

**Comment by alesito85** [comment]
I have issues using method omni_funded_sendall.

For example I have balance 7.7 propertyid 1 on one address and minimum BTC (546e-5). I'm funding from another address that has 5+ BTC balance and plenty of propertyid 1 (even though none is necessary). But whenever I try to transfer I get invalid reason "Sender has no tokens to send".

On top of that the method doesn't work at all if the from address (which has 7.7 - not a funding address) has 0 BTC balance. In that case I get "Error choosing inputs for the send transaction".

**Comment by dexX7** [comment]
I think there is a misunderstanding:

> But whenever I try to transfer I get invalid reason "Sender has no tokens to send".

Tokens are only spent from the very first address.

> On top of that the method doesn't work at all if the from address (which has 7.7 - not a funding address) has 0 BTC balance.

This is expected, because at least some BTC is needed.

**Comment by dexX7** [comment]
I published instructions for the new funding RPCs:

https://medium.com/omnilayer/introducing-funded-omni-layer-transactions-or-how-exchanges-can-save-money-f3f04161a6af
